main.tf

```diff
@@ -68,179 +83,104 @@ resource "aws_s3_bucket" "s3_bucket" {
 }
 
 resource "aws_iam_user" "s3_bucket_iam_user" {
-  name = "${var.bucket_iam_user}"
+  count = "${var.number_of_users}"
+
+  name = "${var.bucket_iam_user}${var.number_of_users != 1 ? "-${count.index}" : "" }"
   path = "/"
 }
 
-resource "aws_iam_user_policy" "s3_bucket_with_kms_user_policy" {
-  count = "${var.kms_alias == "" ? 0 : 1 }"
+resource "aws_iam_user_policy" "s3_bucket_with_kms_user_policy_1" {
+  count = "${var.kms_alias != "" && length(var.white_list_ip) == 0 ? var.number_of_users : 0}"
 
-  name   = "${var.iam_user_policy_name}"
-  user   = "${aws_iam_user.s3_bucket_iam_user.name}"
-  policy = "${data.aws_iam_policy_document.s3_bucket_with_kms_policy_document.json}"
+  name   = "${var.iam_user_policy_name}1"
```

Comment:
In the console these names will be difficult to interpret. I'd suggest updating with additional information based on what the configuration being applied is (e.g. WhitelistIP policy etc)